### ISSUE TITLE
How to do local convolution-like calculations in pseudocolor plot

### ISSUE THREAD

**Issue by markcmiller86** [visit-dav/live-customer-response#158]
We're trying to [combine `visit-users` email list with GitHub issues](https://github.com/visit-dav/live-customer-response/wiki/How-the-new-GitHub-visit-users-Email-Integration-Works).
When replying on `visit-users`, please reply **only** to emails with a *Subject*
line that includes `[visit-dav/live-customer-response]`.

---

I need some help to do convolution-like calculations between two scalars in a 2D mesh. 
What would I like to do?
I have a scalar variable var1 whose 2D pseudocolor plot looks like. 

![image](https://user-images.githubusercontent.com/5720676/95141609-82d3c380-0726-11eb-9a7c-1daf8ac2d735.png)

And I would like to locally convolute this scalar variable var1 with a cylindrically-symmetric 
gaussian function just inside a circular subregion/patch of var1, e.g. 

![image](https://user-images.githubusercontent.com/5720676/95141637-92530c80-0726-11eb-8c80-5efca8dcf715.png)

then move the patch and repeat the process until it sweeps the whole domain. Locally, the gaussian function is always the same. 
And finally, I would like to save all convolution results (of each patch region) in a new scalar variable (if possible) to pseudocolor plot later. 

How am I trying/planning to do it?
Let's give the name var2 to the gaussian function, and var3 to the final result. 
Inside each patch the actual result I'm looking for it's of the type: 

conv=Query("Variable Sum","var1*var2")
DefineScalarExpression("var3","conv")

The main issue with this is that I am not able to evaluate independently the scalar mesh variable var3 for each local region.

So, how could I deal with this? Is there any other best way to do this?

**Comment by markcmiller86** [comment]
So, I wanted to see if I could achieve what you're aiming for with Data-Binning operations in VisIt. I think I can get close but not quite. And, I'd still wind up with an outer-loop iteration having to be performed in Python CLI logic.

Once you need to write python code to do *any* of this, I kinda sorta think it might be the most productive to do *all* of it in Python rather than some hybrid of VisIt expressions, queries, operators and an outer-loop iteration. A guess one question is what *scales* you expect to run these kinds of operations at in terms of grid points? If your example here is about the size you would ever need, then coding *all* operations in Python would definitely be fine. If, however, you might need to run on this on grids with a bazillion points, then we'd probably wanna do the work using a hybrid VisIt+CLI approach and normal python.

**Comment by markcmiller86** [comment]
Well, now that I look further, maybe the best option is a [VisIt Python Expression](https://www.visitusers.org/index.php?title=VisIt-tutorial-Advanced-scripting#Python_Expression_Tutorial)

**Comment by markcmiller86** [comment]
Ok, I have made some progress here with a Python expression. I tested on an input, 2D dataset 50x50 rectilinear grid...

![Screen Shot 2020-10-06 at 11 58 35 AM](https://user-images.githubusercontent.com/5720676/95247857-44dcab00-07cb-11eb-9af0-afc16921d471.png)

And am producing this gaussian-filtered result...

![Screen Shot 2020-10-06 at 11 59 29 AM](https://user-images.githubusercontent.com/5720676/95247935-650c6a00-07cb-11eb-9b3c-f7eca23f15ad.png)

Knowing the dimensions, I can hard-code those into my python expression code...

```
from math import sin, pi, exp, sqrt
class MyExpression(SimplePythonExpression):
    def __init__(self):
        SimplePythonExpression.__init__(self)
        self.name = "MyExpression"
        self.description = "Perturb input scalar by simple wave pattern."
        self.output_is_point_var  = False
        self.output_dimension = 1
        self.kernel_weights = []
    def modify_contract(self,contract):
        pass
    def build_kernel_weights(self, Nx, dx, Ny, dy, sigma):
        # Build just the upper right quadrant
        self.kernel_weights = [[0,]*Nx,]*Ny
        for i in xrange(Nx):
            for j in xrange(Ny):
                x = i*dx
                y = j*dy
                r2 = x*x + y*y
                w = exp(-r2/(2*sigma*sigma))/(sqrt(2*pi)*sigma)
                self.kernel_weights[i][j] = w
    def kernel_weight(self, i, j):
        if i < 0:
            i = -i
        if j < 0:
            j = -j
        return self.kernel_weights[i][j]
    def compute_dxdy(self, ds_in, mode=None):
        if mode == "average":
            ncells = ds_in.GetNumberOfCells()
            dx = dy = 0
            for i in xrange(ncells):
                cell = ds_in.GetCell(i)
                bounds = cell.GetBounds()
                dx += (bounds[1] - bounds[0])
                dy += (bounds[3] - bounds[2])
            dx /= float(ncells)
            dy /= float(ncells)
        else:
            cell0 = ds_in.GetCell(0)
            bounds = cell0.GetBounds()
            dx = bounds[1] - bounds[0]
            dy = bounds[3] - bounds[2]
        return dx, dy
    def derive_variable(self,ds_in,domain_id):
        # ds_in is a vtk dataset, we want
        # to create and return a new vtkDataArray
        # that contains a simple sine wave pattern
        sigma = 1.0
        nx = ny = 5 # 15x15 stencil
        dx, dy = self.compute_dxdy(ds_in)
        self.build_kernel_weights(nx, dx, ny, dy, sigma) 
        cell_vals = ds_in.GetCellData().GetArray(self.input_var_names[0])
        ncells = ds_in.GetNumberOfCells()
        res = vtk.vtkFloatArray()
        res.SetNumberOfComponents(1)
        res.SetNumberOfTuples(ncells)
        for k in xrange(ncells):
            ki = k%50
            kj = k/50
            val = 0.0
            cell = ds_in.GetCell(k)
            bounds = cell.GetBounds()
            xc = bounds[0] + bounds[1] / 2.0
            yc = bounds[2] + bounds[3] / 2.0
            for i in xrange(-(nx-1),nx):
                for j in xrange(-(ny-1),ny):
                    kii = ki + i
                    kjj = kj + j
                    kk = kjj*50+kii
                    pcoords = [0.0,]*3
                    weights = [0.0,]*3
                    x = xc + i*dx
                    y = yc + j*dy
                    subid = vtk.reference(0)
                    w = self.kernel_weight(i,j)
                    if kk>=0 and kk<ncells:
                        val += cell_vals.GetTuple(kk)[0]*w
                    #cid = ds_in.FindCell([xc,yc,0],cell,k,(dx+dy)/10.0,subid,pcoords,weights)
                    #if cid>=0 and cid<=ncells:
                    #   val += cell_vals.GetTuple(cid)[0] * w
            res.SetTuple1(k,val)
        return res
 
py_filter = MyExpression

**Comment by markcmiller86** [comment]
This assymmetry in the above result has me concerned. I suspect it is some kind of off-by-one error somewhere though.

**Comment by markcmiller86-visit** [comment]
Wow..., thanks a lot Mark for your answers!
I suspect your python expression is what I'm looking for. You are pushing
my knowledge's limits.

>>A guess one question is what *scales* you expect to run these kinds of
operations at in terms of grid points?
It's a 2D 300x300 cartesian grid.

To be sure about your lexic:
>>then we'd probably wanna do the work using a hybrid VisIt+CLI approach
and normal python.
May you explain a bit more about the concept of hybrid VisIt+CLI approach
and normal python?
With VisIt+CLI I understand: everything in the visit python interface
manual.

>>Once you need to write python code to do *any* of this, I kinda sorta
think it might be the most productive to do *all* of it in Python rather
than some >>hybrid of VisIt expressions, queries, operators and an
outer-loop iteration
When you said "to do all of it in python" or "normal python", are you
referring to just use VisIt+CLI to read data and call *visit python
expressions,*
and after that continue in python?

**Comment by markcmiller86** [comment]
> May you explain a bit more about the concept of hybrid VisIt+CLI approach and normal python? With VisIt+CLI I understand: everything in the visit python interface manual.

Well, I started thinking it might be easier to use [NumPy](https://medium.com/analytics-vidhya/2d-convolution-using-python-numpy-43442ff5f381) to do this either NumPy alone or maybe VisIt *and* NumPy. But, I've now got something that is close using a Python Expression and so would now recommend doing that.

> When you said "to do all of it in python" or "normal python", are you referring to just use VisIt+CLI to read data and call *visit python expressions,* and after that continue in python?

Well, again, when I started working on an answer, I hadn't considered using Python Expressions. I think Python Expressions are the way to go here though and so you can ignore my earlier remarks about this.

The code, above, is intended to run as a [VisIt Python Expression](https://www.visitusers.org/index.php?title=VisIt-tutorial-Advanced-scripting) so it won't work as stand-alone Python code.

And, when I started writing it, I was trying to write it for the general case of an unstructured mesh. The idea is to run a *stencil* of `nx` X `ny` cell samples centered *around* the input cell and define a set of weights to be used within each stencil. The delta-x and delta-y sample spacing can be computed either by assuming a uniform grid and using the first cell in the grid as the spacing or computing an average cell size in X and Y and using that (`def compute_dxdy()` method in the expression). In my case, I define the weights using a radially symmetric, Gaussian filter defined by (defined by `def build_kernel_weights()`). Then, I intended to just iterate over all cells in the mesh, using the cell center, `xc` and `yc` and the i/j offets within the kernel to *find* cells associated with those sample points (`ds_in.FindCell()` commented out). That approach failed because `FindCell()` didn't seem to work as I expected. So, I hard-coded the logic to assume a uniform, structured mesh of 50 x 50 and that seems to have worked much better. I am hoping you can, of course, adjust it to use different filter weights and different mesh size (300 x 300).

**Comment by markcmiller86-visit** [comment]
It's much clearer now.
I have a lot of work to do here.
I really appreciate you have taken the time!

**Comment by markcmiller86** [comment]
Ok, I've found all my bugs in the above and have this working. It no longer assumes structured grid and you don't have to know dimensions either...

Input...

![Screen Shot 2020-10-06 at 4 28 21 PM](https://user-images.githubusercontent.com/5720676/95270700-f510da80-07f0-11eb-93cc-4c32124276df.png)

Output (using a simple box filter (1/((2*nx-1)*(2*ny-1))

![Screen Shot 2020-10-06 at 4 28 06 PM](https://user-images.githubusercontent.com/5720676/95270687-ecb89f80-07f0-11eb-8df0-749203a1c1fe.png)

Output using the radial Gaussian filter...

![Screen Shot 2020-10-06 at 4 31 32 PM](https://user-images.githubusercontent.com/5720676/95270903-65b7f700-07f1-11eb-934e-08e0434ef03c.png)

Here is the corrected code...

```
from math import sin, pi, exp, sqrt
class MyExpression(SimplePythonExpression):
    def __init__(self):
        SimplePythonExpression.__init__(self)
        self.name = "MyExpression"
        self.description = "Perform convolution with box or radial gaussian"
        self.output_is_point_var  = False
        self.output_dimension = 1
        self.kernel_weights = []
    def modify_contract(self,contract):
        pass
    def build_kernel_weights(self, Nx, dx, Ny, dy, sigma):
        # Build just the upper right quadrant
        self.kernel_weights = [[0,]*Nx,]*Ny
        for i in xrange(Nx):
            for j in xrange(Ny):
                x = i*dx
                y = j*dy
                r2 = x*x + y*y
                w = exp(-r2/(2*sigma*sigma))/(sqrt(2*pi)*sigma)
                self.kernel_weights[i][j] = w
    def build_box_kernel_weights(self, nx, dx, ny, dy):
        self.kernel_weights = [[0,]*nx,]*ny
        for i in xrange(nx):
            for j in xrange(ny):
                self.kernel_weights[i][j] = 1.0/((2*nx-1)*(2*ny-1))
    def kernel_weight(self, i, j):
        if i < 0:
            i = -i
        if j < 0:
            j = -j
        return self.kernel_weights[i][j]
    def compute_dxdy(self, ds_in, mode=None):
        if mode == "average":
            ncells = ds_in.GetNumberOfCells()
            dx = dy = 0
            for i in xrange(ncells):
                cell = ds_in.GetCell(i)
                bounds = cell.GetBounds()
                dx += (bounds[1] - bounds[0])
                dy += (bounds[3] - bounds[2])
            dx /= float(ncells)
            dy /= float(ncells)
        else:
            cell0 = ds_in.GetCell(0)
            bounds = cell0.GetBounds()
            dx = bounds[1] - bounds[0]
            dy = bounds[3] - bounds[2]
        return dx, dy
    def derive_variable(self,ds_in,domain_id):
        sigma = 1.0
        nx = ny = 4 # 2nx-1 x 2ny-1 stencil
        dx, dy = self.compute_dxdy(ds_in)
        self.build_kernel_weights(nx, dx, ny, dy, sigma) 
        cell_vals = ds_in.GetCellData().GetArray(self.input_var_names[0])
        ncells = ds_in.GetNumberOfCells()
        res = vtk.vtkFloatArray()
        res.SetNumberOfComponents(1)
        res.SetNumberOfTuples(ncells)
        for k in xrange(ncells):
            val = 0
            cell = ds_in.GetCell(k)
            bounds = cell.GetBounds()
            xc = (bounds[0] + bounds[1]) / 2.0
            yc = (bounds[2] + bounds[3]) / 2.0
            for i in xrange(-(nx-1),nx):
                for j in xrange(-(ny-1),ny):
                    pcoords = [0.0,]*3
                    weights = [0.0,]*3
                    x = xc + i*dx
                    y = yc + j*dy
                    subid = vtk.reference(0)
                    w = self.kernel_weight(i,j)
                    cid = ds_in.FindCell([x,y,0],cell,k,(dx+dy)/10.0,subid,pcoords,weights)
                    if cid>=0 and cid<ncells:
                       val += cell_vals.GetTuple1(cid) * w
                    else:
                        val += w
            res.SetTuple1(k,val)
        return res
 
py_filter = MyExpression
```

**Comment by markcmiller86** [comment]
@cyrush ... a few quick questions...

- Is this example (or suitable refinement thereof) appropriate for docs?
- In the case of multi-block, parallel decomposed input, is there a way to handle the possible 2nd, 3rd, 4th or more layer of ghost zoning? Or, is there a way to handle the MPI comm here in the Python filter code that would get the right answer for multi-block case?
- Would it make sense to productize a general stencil filter and make it part of VisIt so that users could easily perform same but create whatever filter kernel weights and shapes they wanted?

**Comment by brugger1** [comment]
I think this is definitely suitable for our documentation. I'm not sure if it's appropriate for the analysis section of the manual or if we could have a section on python filter examples.

**Comment by markcmiller86-visit** [comment]
@marck...!
Now I understand what (and how) we are doing here.

>>Well, now that I look further, maybe the best option is a VisIt Python
Expression
<https://www.visitusers.org/index.php?title=VisIt-tutorial-Advanced-scripting#Python_Expression_Tutorial>
I began running this tutorial visit python expression script and everything
went well.

Next I moved to your script, and a problem appeared in line 75,
subid = vtk.reference(0)

The error message is:

VisIt: Error - Pseudocolor:  ()
viewer: The 'ases' expression failed because avtPythonExpression::Execute
Error - Python Expression 'execute' method failed
Python Environment Error:
<type 'exceptions.AttributeError'>'module' object has no attribute
'reference'
Traceback (most recent call last):
  File
"/home/ernesto/programas/visit2_13_2.linux-x86_64/2.13.2/linux-x86_64/lib/site-packages/pyavt/filters.py",
line 101, in execute
    res = self.derive_variable(data_sets[i],domain_ids[i])
  File "<string>", line 75, in derive_variable
AttributeError: 'module' object has no attribute 'reference'

Could it be that I need to install any (vtk) library or any header file is
missing?
